moto/route53domains/models.py

```diff
@@ -0,0 +1,309 @@
+                [f"Domain {domain_name} isn't registered in the current account"]
+            )
+
+        self.__operations[requested_operation.id] = requested_operation
+        del self.__domains[domain_name]
+        return requested_operation
+
+    def __validate_duplicate_operations(
+        self, requested_operation: Route53DomainsOperation
+    ):
+        for operation in self.__operations.values():
+            if (
+                operation.domain_name == requested_operation.domain_name
+                and operation.type == requested_operation.type
+            ):
+                raise DuplicateRequestException()
+
+    def get_domain(self, domain_name: str) -> Route53Domain:
+        input_errors: List[str] = []
+        Route53Domain.validate_domain_name(domain_name, input_errors)
+        if input_errors:
+            raise InvalidInputException(input_errors)
+
+        if domain_name not in self.__domains:
+            raise InvalidInputException(["Domain is not associated with this account"])
+
+        return self.__domains[domain_name]
+
+    def list_domains(
+        self,
+        filter_conditions: Optional[List[Dict]] = None,
+        sort_condition: Optional[Dict] = None,
+        marker: Optional[str] = None,
+        max_items: Optional[int] = None,
+    ) -> Tuple[List[Route53Domain], Optional[str]]:
+        try:
+            filters: List[DomainsFilter] = (
+                [DomainsFilter.validate_dict(f) for f in filter_conditions]
+                if filter_conditions
+                else []
+            )
+            sort: Optional[DomainsSortCondition] = (
+                DomainsSortCondition.validate_dict(sort_condition)
+                if sort_condition
+                else None
+            )
+        except ValidationException as e:
+            raise InvalidInputException(e.errors)
+
+        filter_fields = [f.name for f in filters]
+        if sort and filter_fields and sort.name not in filter_fields:
+            raise InvalidInputException(
+                ["Sort condition must be the same as the filter condition"]
+            )
+
+        domains_to_return = []
+        max_items = max_items or self.DEFAULT_MAX_ITEMS
+
+        for domain in self.__domains.values():
+            if len(domains_to_return) == max_items:
+                break
+
+            if all([f.filter(domain) for f in filters]):
+                domains_to_return.append(domain)
+
+        if sort:
+            if sort.name == DomainFilterField.domain_name:
+                domains_to_return.sort(
+                    key=lambda d: d.domain_name,
+                    reverse=(sort.sort_order == DomainSortOrder.descending),
+                )
+            else:
+                domains_to_return.sort(
+                    key=lambda d: d.expiration_date,
+                    reverse=(sort.sort_order == DomainSortOrder.descending),
+                )
+
+        start_idx = 0 if marker is None else int(marker)
```

Comment:
The pagination can probably be handled by the `@paginate`-decorator. That would simplify the logic, and give you a lot of validation out of the box.
See the documentation here: http://docs.getmoto.org/en/latest/docs/contributing/development_tips/utilities.html#paginator

I think that the docs has all the information you need, but let me know if run into any problems.

Reply:
Awesome! I changed the functions as per your suggestion and it worked. I also removed the test for testing the pagination logic since it seems out of the scope of the test module.

I really liked the pagination utility, really neat solution!

Reply:
Awesome! I changed the functions as per your suggestion and it worked. I also removed the test for testing the pagination logic since it seems out of the scope of the test module.

I really liked the pagination utility, really neat solution!